### PR TITLE
Placeholder for clearing session cache for unjailed vals with upgrade height

### PR DESCRIPTION
**Pull request by nodiesBlade**
Disclaimer: Never done an upgrade height transaction before, please do a test in local net, testnet, and so on for this upgrade

Addresses concerns in #1529 by adding an upgrade height

### CODE REVIEW

File: x/gov/module.go [hunk 1]
```diff
@@ -120,6 +121,12 @@ func (am AppModule) BeginBlock(ctx sdk.Ctx, req abci.RequestBeginBlock) {
 
 	ActivateAdditionalParametersACL(ctx, am)
 
+	// On this upgrade height, Pocket Core will start clearing the global session cache to prevent any non-deterministic cache consistency issues.
+	// This code will ensure it is cleared on the upgrade height for a fresh start.
+	if am.keeper.GetCodec().IsOnNamedFeatureActivationHeight(ctx.BlockHeight(), codec.ClearUnjailedValSessionKey) {
```
Comment (Olshansk) [inline x/gov/module.go]: I'd probably add some tolerance (+-2 blocks) just to be safe. 

Whatever hit we take in performance during these few blocks in `BeginBlock` isn't worth the lack of testing/infrastructure we have to be confident that no caching issue will break anything.

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Comment by Olshansk** [comment]
@msmania PTAL when you have a chance as another pair of 👀

File: x/gov/module.go (hunk 1 above)
Comment (nodiesBlade) [reply x/gov/module.go]: I'm not sure if this is necessary to clear the cache so many times, but regardless added it along with some unit tests.

**Comment by msmania** [comment]
Sorry for being late, and thank you @PoktBlade for creating this.  I totally agree with your comment https://github.com/pokt-network/pocket-core/issues/1529#issuecomment-1459082393.  It's technically possible that the network reaches wrong consensus if the majority of validators end up having unupdated session cache.

This patch looks good to me.

File: x/nodes/keeper/valStateChanges.go [hunk 2]
```diff
@@ -652,6 +652,12 @@ func (k Keeper) UnjailValidator(ctx sdk.Ctx, addr sdk.Address) {
 		k.Logger(ctx).Error(fmt.Sprintf("cannot unjail already unjailed validator, validator: %v at height %d\n", validator, ctx.BlockHeight()))
 		return
 	}
+
+	// After or on this upgrade height, we will start clearing the global session cache to prevent any non-deterministic cache consistency issues.
+	if k.Cdc.IsAfterNamedFeatureActivationHeight(ctx.BlockHeight(), codec.ClearUnjailedValSessionKey) {
```
Comment (Olshansk) [inline x/nodes/keeper/valStateChanges.go:657]:
```suggestion
	// See #1529 for more details
	if k.Cdc.IsAfterNamedFeatureActivationHeight(ctx.BlockHeight(), codec.ClearUnjailedValSessionKey) {
```

File: x/gov/module.go [hunk 3]
```diff
@@ -161,6 +168,7 @@ func ActivateAdditionalParametersACL(ctx sdk.Ctx, am AppModule) {
 		params.ACL.SetOwner(types.NewACLKey(types.NodesSubspace, "ServicerStakeFloorMultiplierExponent"), am.keeper.GetDAOOwner(ctx))
 		am.keeper.SetParams(ctx, params)
 	}
+
```
Comment (Olshansk) [inline x/gov/module.go:171]:
```suggestion
```

File: codec/codec.go [hunk 4]
```diff
@@ -281,6 +282,16 @@ func (cdc *Codec) IsOnNamedFeatureActivationHeight(height int64, key string) boo
 	return UpgradeFeatureMap[key] != 0 && height == UpgradeFeatureMap[key]
 }
 
+func (cdc *Codec) IsBetweenNamedFeatureActivationHeight(height int64, key string, tolerance int64) bool {
```
Comment (Olshansk) [inline codec/codec.go:285]:
```suggestion
// IsOnNamedFeatureActivationHeightWithTolerance is used to enable certain business logic within some tolerance
// (i.e. only a few blocks) of feature activation to have more confidence in the feature's release and avoid
// non-deterministic or hard-to-predict behaviour.
func (cdc *Codec) IsOnNamedFeatureActivationHeightWithTolerance(height int64, featureKey string, tolerance int64) bool {
```

Comment (Olshansk) [reply codec/codec.go:285]: Make sure to change the test name et al too

File: codec/codec_test.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestIsBetweenNamedFeatureActivationHeight(t *testing.T) {
+	tolerance := int64(5)
+	codecUpgradeKey := ClearUnjailedValSessionKey
+
+	// have to init global map / not mock friendly.
```
Comment (Olshansk) [inline codec/codec_test.go:12]: 1. Add a comment at the top that this should never be run in parallel
2. Capture the value of `UpgradeFeatureMap` before the test
3. Run `t.Cleanup(...)` where you reassign the value it had pre-test

File: x/gov/module.go [hunk 6]
```diff
@@ -120,6 +121,12 @@ func (am AppModule) BeginBlock(ctx sdk.Ctx, req abci.RequestBeginBlock) {
 
 	ActivateAdditionalParametersACL(ctx, am)
 
+	// On this upgrade height, Pocket Core will start clearing the global session cache to prevent any non-deterministic cache consistency issues.
```
Comment (Olshansk) [inline x/gov/module.go:124]: s/On/Around

File: x/gov/module.go [hunk 7]
```diff
@@ -120,6 +121,12 @@ func (am AppModule) BeginBlock(ctx sdk.Ctx, req abci.RequestBeginBlock) {
...
+	// This code will ensure it is cleared on the upgrade height for a fresh start.
+	if am.keeper.GetCodec().IsBetweenNamedFeatureActivationHeight(ctx.BlockHeight(), codec.ClearUnjailedValSessionKey, 2) {
```
Comment (Olshansk) [inline x/gov/module.go:126]: move magic number into a const with a descriptive name

**Review by Olshansk — CHANGES_REQUESTED** [review]

**Comment by Olshansk** [comment]
@PoktBlade Okay to close this one out in favor of #1529 so we can finish it off there?

**Comment by Olshansk** [comment]
Closing it seeing how #1529 was merged in.

**Pull request closed without merging**